barretenberg/cpp/src/barretenberg/smt_verification/terms/term.cpp

```diff
@@ -0,0 +1,468 @@
+        case TermType::FFTerm:
+            this->term = slv->term_manager.mkFiniteFieldElem(t, slv->ff_sort, base);
+            break;
+        case TermType::FFITerm:
+            // TODO(alex): CVC5 doesn't provide integer initialization from hex. Yet.
+            strvalue = slv->term_manager.mkFiniteFieldElem(t, slv->ff_sort, base).getFiniteFieldValue();
+            this->term = slv->term_manager.mkInteger(strvalue);
+            this->mod();
+            break;
+        case TermType::BVTerm:
+            // it's better to have (-p/2, p/2) representation due to overflows
+            strvalue = slv->term_manager.mkFiniteFieldElem(t, slv->ff_sort, base).getFiniteFieldValue();
+            this->term = slv->term_manager.mkBitVector(slv->bv_sort.getBitVectorSize(), strvalue, 10);
+            break;
+        }
+    }
+}
+
+void STerm::mod()
+{
+    if (this->type == TermType::FFITerm) {
+        cvc5::Term modulus = this->solver->term_manager.mkInteger(solver->modulus);
+        this->term = this->solver->term_manager.mkTerm(this->operations.at(OpType::MOD), { this->term, modulus });
+    }
+}
+
+STerm STerm::operator+(const STerm& other) const
+{
+    cvc5::Term res = this->solver->term_manager.mkTerm(this->operations.at(OpType::ADD), { this->term, other.term });
+    return { res, this->solver, this->type };
+}
+
+void STerm::operator+=(const STerm& other)
+{
+    this->term = this->solver->term_manager.mkTerm(this->operations.at(OpType::ADD), { this->term, other.term });
+}
+
+STerm STerm::operator-(const STerm& other) const
+{
+    cvc5::Term res = this->solver->term_manager.mkTerm(this->operations.at(OpType::NEG), { other.term });
+    res = solver->term_manager.mkTerm(this->operations.at(OpType::ADD), { this->term, res });
+    return { res, this->solver, this->type };
+}
+
+void STerm::operator-=(const STerm& other)
+{
+    cvc5::Term tmp_term = this->solver->term_manager.mkTerm(this->operations.at(OpType::NEG), { other.term });
+    this->term = this->solver->term_manager.mkTerm(cvc5::Kind::FINITE_FIELD_ADD, { this->term, tmp_term });
+}
+
+STerm STerm::operator-() const
+{
+    cvc5::Term res = this->solver->term_manager.mkTerm(this->operations.at(OpType::NEG), { this->term });
+    return { res, this->solver, this->type };
+}
+
+STerm STerm::operator*(const STerm& other) const
+{
+    cvc5::Term res = solver->term_manager.mkTerm(this->operations.at(OpType::MUL), { this->term, other.term });
+    return { res, this->solver, this->type };
+}
+
+void STerm::operator*=(const STerm& other)
+{
+    this->term = this->solver->term_manager.mkTerm(this->operations.at(OpType::MUL), { this->term, other.term });
+}
+
+/**
+ * @brief Division operation
+ *
+ * @details Returns a result of the division by
+ * creating a new symbolic variable and adding a new constraint
+ * to the solver.
+ *
+ * @param other
+ * @return STerm
+ */
+STerm STerm::operator/(const STerm& other) const
+{
+    if (!this->operations.contains(OpType::DIV)) {
```

Comment:
Uhm. Why can't you replace it with multilication and non-zero?

Reply:
Because of BitVectors. They are not supposed to be divided by each other(well at least not like the real division, truncated - maybe. At least for now they are not)